transformation/transformationPlugins/jsonMessageTransformers/jsonMessageTransformerInterface/src/main/java/org/opensearch/migrations/transform/ThreadSafeTransformerWrapper.java

```diff
@@ -0,0 +1,132 @@
+package org.opensearch.migrations.transform;
+
+import java.lang.ref.Cleaner;
+import java.util.function.Supplier;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A thread-safe wrapper around {@link IJsonTransformer} that ensures each thread
+ * has its own transformer instance via {@link ThreadLocal}.
+ * <p>
+ * This design is useful when the transformer is stateful or not thread-safe.
+ * Each thread gets a lazily-initialized transformer from the provided supplier.
+ *
+ * <h2>Resource Management</h2>
+ * The wrapper implements {@link AutoCloseable}, and callers are expected to invoke
+ * {@link #close()} when a thread is finished using its transformer. This ensures
+ * deterministic release of resources and avoids memory leaks in thread pools.
+ *
+ * <h2>Cleaner Integration</h2>
+ * As a fallback, this class also registers each transformer with a shared {@link java.lang.ref.Cleaner}.
+ * If a thread dies or the transformer becomes unreachable without an explicit {@link #close()} call,
+ * the cleaner will invoke a cleanup task to call {@link IJsonTransformer#close()}.
+ *
+ * <h2>Thread-local semantics</h2>
+ * Calling {@link #close()} only affects the transformer instance associated with the
+ * current thread. It does not close transformers used by other threads.
+ */
+@Slf4j
+public class ThreadSafeTransformerWrapper implements IJsonTransformer {
+
+    public static final Cleaner FALLBACK_CLEANER = Cleaner.create();
+
+    /** Thread-local holding per-thread transformer wrappers */
+    private final ThreadLocal<CloseTrackingTransformer> threadLocalHolder;
```

Comment:
nit: holder is a little bit weird.  Is this your main cache of transformers or just ancillary cleanup related stuff?